README.md
# MCTranslator

为 Minecraft Mod, 插件 等翻译工作提供帮助

在使用中如果遇到问题, 或者你有一些意见或建议 欢迎通过 issues 或 [电子邮箱](mailto:Hum-XC@outlook.com) 与我交流。

## 使用演示

![使用演示](https://files.catbox.moe/n9fyva.gif)

## 使用方法

1、前往 [百度翻译开放平台](https://fanyi-api.baidu.com/) 可以申请免费到免费的 APPID 和 密钥，填到插件对应的配置项中即可开始使用。

2、在资源浏览器或者打开的文件窗口中点击右键，再点击 \[翻译语言文件\] 选项。在新打开的窗口中，你可以浏览，修改给出的翻译内容。

3、每个翻译条目的右边都有一个选择框，在调整到你满意的结果之后，选中你需要保存的条目，然后点击保存，就会保存成果到目录下一个以 "tr" 为前缀的文件。

## 自动翻译

插件默认开启\[自动翻译成中文\], 即在按下\[翻译语言文件\]后自动翻译成中文, 你仍然可以在翻译工作台页面修改翻译语言。

## 修改翻译语言

在翻译工作台页面顶部可以设置翻译语言, 输入框内填写的是语言代码, 下面是常用的语言代码。更多可以在[这里](https://fanyi-api.baidu.com/doc/21)找到。

| 名称         | 代码 | 名称       | 代码 | 名称       | 代码 |
| ------------ | ---- | ---------- | ---- | ---------- | ---- |
| 自动检测     | auto | 中文       | zh   | 英语       | en   |
| 粤语         | yue  | 文言文     | wyw  | 日语       | jp   |
| 韩语         | kor  | 法语       | fra  | 西班牙语   | spa  |
| 泰语         | th   | 阿拉伯语   | ara  | 俄语       | ru   |
| 葡萄牙语     | pt   | 德语       | de   | 意大利语   | it   |
| 希腊语       | el   | 荷兰语     | nl   | 波兰语     | pl   |
| 保加利亚语   | bul  | 爱沙尼亚语 | est  | 丹麦语     | dan  |
| 芬兰语       | fin  | 捷克语     | cs   | 罗马尼亚语 | rom  |
| 斯洛文尼亚语 | slo  | 瑞典语     | swe  | 匈牙利语   | hu   |
| 繁体中文     | cht  | 越南语     | vie  |            |      |

## 注意事项

1、在一些翻译条目中会存在特殊符号，比如 \[ \n , $ , & \]。而在结果翻译之后，他们可能会错位, 或者结果不正确, 比如某些符号会被翻译成中文符号。这些符号对 游戏 mod 来说可能很重要，所以请仔细观察他们，手动修正这些错误的符号。在翻译工作台页面，一些重要的符号会被显示成红色。


CHANGELOG.md
<!--
 * @Author: HumXC Hum-XC@outlook.com
 * @Date: 2022-05-24
 * @LastEditors: HumXC Hum-XC@outlook.com
 * @LastEditTime: 2022-07-26
 * @FilePath: \MCTranslator\CHANGELOG.md
 * @Description:
 *
 * Copyright (c) 2022 by HumXC Hum-XC@outlook.com, All Rights Reserved.
-->

# Change Log

## [Unreleased]

-   Initial release

## [0.0.1] - 2022-05-25

### Added

-   完成基本功能

## [0.0.2] - 2022-05-26

### Changed

-   美化页面

## [0.0.3] - 2022-05-26

### Added

-   增加翻译语言选择

## [0.0.4] - 2022-05-27

### Added

-   给条目选择框添加序号

### Fixed

-   修复译文中出现特殊符号时会破坏页面和导致翻译异常的问题

## [0.0.5] - 2022-07-26

### Added

-   高级版 API 配置项

### Fixed

-   修复译文为空字符时造成的异常


src/util.ts
/*
 * @Author: HumXC Hum-XC@outlook.com
 * @Date: 2022-05-25
 * @LastEditors: HumXC Hum-XC@outlook.com
 * @LastEditTime: 2022-07-26
 * @FilePath: \MCTranslator\src\util.ts
 * @Description:
 *
 * Copyright (c) 2022 by HumXC Hum-XC@outlook.com, All Rights Reserved.
 */
import vscode = require("vscode");
import path = require("path");
const { readFile } = require("fs").promises;

export const readHtml = async (htmlPath: string, panel: vscode.WebviewPanel) =>
    (await readFile(htmlPath, "utf-8"))
        .replace(/%CSP_SOURCE%/gu, panel.webview.cspSource)
        .replace(
            /(src|href)="([^"]*)"/gu,
            (_: any, type: any, src: string) =>
                `${type}="${panel.webview.asWebviewUri(
                    vscode.Uri.file(path.resolve(htmlPath, "..", src))
                )}"`
        );
export function sleep(timeout: number): Promise<void> {
    return new Promise<void>((resolve) => {
        setTimeout(resolve, timeout);
    });
}


src/network.ts
import * as https from "https";
import * as crypto from "crypto";
var baiduError = new Map<number, string>([
    [52000, "成功"],
    [52001, "请求超时，请重试"],
    [52002, "系统错误，请重试"],
    [52003, "未授权用户，请检查appid是否正确或者服务是否开通"],
    [54000, "必填参数为空，请检查是否少传参数"],
    [54001, "签名错误，请检查您的签名生成方法"],
    [54003, "访问频率受限，请降低您的调用频率，或进行身份认证后切换为高级版/尊享版"],
    [54004, "账户余额不足，请前往管理控制台为账户充值"],
    [54005, "长query请求频繁，请降低长query的发送频率，3s后再试"],
    [58000, "客户端IP非法，检查个人资料里填写的IP地址是否正确，可前往开发者信息-基本信息修改 "],
    [58001, "译文语言方向不支持，检查译文语言是否在语言列表里"],
    [58002, "服务当前已关闭，请前往管理控制台开启服务"],
    [90107, "认证未通过或未生效，请前往我的认证查看认证进度"],
]);

// 百度翻译请求参数
export type BaiduReq = {
    // 请求翻译query
    q: string;
    // 翻译源语言
    from: string;
    // 翻译目标语言
    to: string;
    // APPID
    appid: string;
    // 随机数
    salt: string;
    // 签名
    sign: string;
};
// 百度翻译回复参数
export type BaiduResp = {
    from: string;
    to: string;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    trans_result: Array<{ src: string; dst: string }>;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    error_code: string;
};
/** 发送http请求 */
function doHttp(options: https.RequestOptions, data: any = undefined): any {
    let a = "";
    return new Promise((resolve, reject) => {
        var respData: string = "";
        const req = https.request(options, (res) => {
            if (res.statusCode !== 200) {
                reject(new Error(res.statusCode?.toString()));
            }
            res.on("data", (d: Buffer) => {
                respData += d.toString("utf-8");
            });
        });
        req.once("error", (error) => {
            reject(error);
        });
        req.once("close", () => {
            try {
                let jsonData = JSON.parse(respData);
                resolve(jsonData);
            } catch (error) {
                reject(error);
            }
        });
        if (data !== undefined) {
            req.write(data);
        }
        req.end();
    });
}
export function baidu(
    str: string,
    from: string,
    to: string,
    appid: string,
    key: string
): Promise<BaiduResp> {
    let salt = Math.ceil(Math.random() * 100000).toString();
    let sign = md5(appid + str + salt + key);
    let data: BaiduReq = {
        q: encodeURI(str),
        from: from,
        to: to,
        appid: appid,
        salt: salt,
        sign: sign,
    };
    let query: string = objToQuery(data);
    let options: https.RequestOptions = {
        hostname: `fanyi-api.baidu.com`,
        path: `/api/trans/vip/translate?` + query,
        method: "GET",
        timeout: 10000,
    };
    return doHttp(options);
}
function md5(str: string): string {
    return crypto.createHash("md5").update(str).digest("hex");
}
// 解析百度api的错误
export function checkBaiduError(resp: BaiduResp): string | undefined {
    let code = Number.parseInt(resp.error_code);
    if (code === 52000) {
        return undefined;
    }
    return baiduError.get(code);
}

// 对象转 GET 使用的 QueryString
function objToQuery(obj: any): string {
    let result = "&";
    for (const key in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, key)) {
            result += `&${key}=${obj[key]}`;
        }
    }
    // 返回时去除开头的 &&
    return result.replace("&&", "");
}


src/extension.ts
import path = require("path");
import * as vscode from "vscode";
import * as fs from "fs";
import { readHtml } from "./util";
import { Document, trHandleBaidu, WebviewDocument } from "./translate";
let outputChannel: vscode.OutputChannel;
export function activate(context: vscode.ExtensionContext) {
    if (outputChannel === undefined) {
        outputChannel = vscode.window.createOutputChannel("MCTranslator");
    }
    output("激活扩展");
    let disposable = vscode.commands.registerCommand("mctranslator.openFile", (uri) => {
        createPanel(context).then((panel: vscode.WebviewPanel) => {
            runCommand(panel, uri);
        });
    });

    context.subscriptions.push(disposable);
}
async function createPanel(context: vscode.ExtensionContext): Promise<vscode.WebviewPanel> {
    const panel = vscode.window.createWebviewPanel(
        "mctranslator.tr",
        "翻译工作台",
        { viewColumn: vscode.ViewColumn.Active, preserveFocus: true },
        {
            enableScripts: true,
            localResourceRoots: [vscode.Uri.file(context.extensionPath)],
            retainContextWhenHidden: true,
        }
    );
    panel.webview.html = await readHtml(
        path.resolve(context.extensionPath, "webview/index.html"),
        panel
    );

    return panel;
}

async function runCommand(panel: vscode.WebviewPanel, uri: vscode.Uri) {
    output("正在启动翻译");
    // 加载的字符串
    let doc: Document = {
        str: "",
        obj: {},
        doctype: "unknow",
        result: [],
        path: uri.fsPath,
    };
    // 获取配置
    let conf = vscode.workspace.getConfiguration("mctranslator");
    let appid = conf.get("百度翻译_AppID") as string;
    let secretkey = conf.get("百度翻译_SecretKey") as string;
    let isAdvancedVersion = conf.get("百度翻译_高级版API") as boolean;
    let autoTrslante = conf.get("自动翻译") as boolean;
    if (appid === undefined || secretkey === undefined) {
        showError("百度翻译配置不正确");
        return;
    }

    // 设置页面
    panel.webview.onDidReceiveMessage(async ({ type, data }) => {
        switch (type) {
            case "save":
                try {
                    let doc: WebviewDocument = JSON.parse(data);
                    let fileName = path.join(
                        path.dirname(doc.path),
                        "tr_" + path.basename(doc.path)
                    );
                    if (doc.doctype === "json") {
                        for (let i = 0; i < doc.result.length; i++) {
                            if (Object.prototype.hasOwnProperty.call(doc.obj, doc.result[i].name)) {
                                doc.obj[doc.result[i].name] = doc.result[i].dst;
                            }
                        }

                        fs.writeFileSync(
                            fileName,
                            JSON.stringify(doc.obj).replace(/","/g, '",\n"')
                        );
                    } else {
                        for (let i = 0; i < doc.result.length; i++) {
                            doc.str = doc.str.replace(
                                new RegExp(`(?<=${doc.result[i].name}\\=).+`, "g"),
                                doc.result[i].dst
                            );
                        }
                        fs.writeFileSync(fileName, doc.str);
                    }
                } catch (error) {
                    panel.webview.postMessage({ type: "SaveFail" });
                    showError("保存失败", error);
                    return;
                }
                panel.webview.postMessage({ type: "Savesuccess" });
                break;

            case "Translate":
                try {
                    let setting = JSON.parse(data);
                    output(
                        `开始翻译(webview): path=${doc.path}\n lang:${setting[0]}->${setting[1]}\n appid=${appid}\n secretkey=${secretkey}`
                    );
                    await trHandleBaidu(
                        doc,
                        setting[0],
                        setting[1],
                        appid,
                        secretkey,
                        isAdvancedVersion
                    );
                } catch (error) {
                    panel.webview.postMessage({ type: "TrslateFail" });
                    showError("请求 API 时出现错误", error);
                    return;
                }
                panel.webview.postMessage({ type: "Loadcomplete", data: JSON.stringify(doc) });
                break;
            case "warn":
                vscode.window.showWarningMessage(data);
                break;
        }
    });
    // 读取文件
    try {
        let d = fs.readFileSync(doc.path);
        doc.str = d.toString("utf-8");
    } catch (error) {
        showError("读取文件时出现了错误", error);
        return;
    }

    doc.doctype = path.extname(uri.fsPath).replace(".", "");
    if (doc.doctype === "json") {
        try {
            doc.obj = JSON.parse(doc.str);
        } catch (error) {
            showError("解析 Json 时出现了错误", error);
        }

        for (const key in doc.obj) {
            if (Object.prototype.hasOwnProperty.call(doc.obj, key)) {
                doc.result.push({ name: key, src: doc.obj[key] as string, dst: "" });
            }
        }
    } else {
        let strArr = doc.str.split("\n");
        for (let i = 0; i < strArr.length; i++) {
            let sp = strArr[i].split("=");
            if (sp[0] === undefined || sp[1] === undefined) {
                continue;
            }
            doc.result.push({ name: sp[0], src: sp[1], dst: "" });
        }
    }
    // 开始自动翻译
    if (autoTrslante) {
        output(
            `开始自动翻译: path=${doc.path}\n lang:auto->zh\n appid=${appid}\n secretkey=${secretkey}`
        );
        try {
            await trHandleBaidu(doc, "auto", "zh", appid, secretkey, isAdvancedVersion);
        } catch (error) {
            showError("请求 API 时出现错误", error);
            return;
        }
    }

    panel.webview.postMessage({ type: "Loadcomplete", data: JSON.stringify(doc) });
}

export function deactivate() {
    output("停止活动");
}
function showError(msg: string, error: any = undefined) {
    let windowMmessage = "";
    let channelMessage = "";
    if (error === undefined || error.message === undefined) {
        windowMmessage = msg;
        channelMessage = msg;
    } else {
        windowMmessage = msg + ": " + (error as Error).message;
        channelMessage = msg + ": \n" + (error as Error).message + "\n" + (error as Error).stack;
    }
    output(channelMessage);
    vscode.window.showErrorMessage(windowMmessage);
}

// 输出到容器
function output(msg: string) {
    outputChannel.appendLine(msg);
}


webview/js/index.js
const vscode = acquireVsCodeApi();
var selectCount;
var checkAll;
var log;
var items;
var doc;
var progress = [];
var $ = (id) => {
    return document.getElementById(id);
};

window.addEventListener("message", (event) => {
    let message = event.data;

    let data;
    try {
        data = JSON.parse(message.data);
    } catch (error) {
        vscode.postMessage({ type: "warn", data: error });
    }

    switch (message.type) {
        case "Loadcomplete":
            doc = data;
            addItem(data.result);
            break;
        case "Savesuccess":
            changeLog("👍保存成功  ");
            break;
        case "SaveFail":
            changeLog("😫保存失败  ");
            break;
        case "TrslateFail":
            changeLog("😫翻译失败  ");
            break;
        default:
            break;
    }
});
function test() {
    $ = (id) => {
        return document.getElementById(id);
    };
    let data = [];
    for (let i = 0; i < 20; i++) {
        data.push({
            name: "a",
            src: "AAsssssssssssssssssssssssssssssssssssssssssssssss",
            dst: "BBBsssssssssssssssssssssssssssssssssss",
        });
    }
    addItem(data);
}
function save(button) {
    doc.result = [];
    for (let i = 0; i < items.length; i++) {
        let _name = items[i].childNodes[5].childNodes[3].title;
        let _dst = items[i].childNodes[5].childNodes[3].value.replace(/\\n/g, "\n");
        if (!items[i].childNodes[3].checked || _dst === "") {
            continue;
        }
        doc.result.push({ name: _name, dst: _dst });
    }
    if (doc.result.length === 0) {
        changeLog("⚠️您尚未选择任何内容  ");
        return;
    }
    logProgress("🤔正在保存  ");
    button.disabled = true;
    setTimeout(() => {
        button.disabled = false;
    }, 3000);
    vscode.postMessage({ type: "save", data: JSON.stringify(doc) });
}
function selectAll(checkAll) {
    if (checkAll.checked) {
        selectCount = items.length;
    } else {
        selectCount = 0;
    }
    changeLog(`${selectCount} / ${items.length}`);
    for (let i = 0; i < items.length; i++) {
        items[i].childNodes[3].checked = checkAll.checked;
    }
    checkAll.className = "checkbox";
}
function addItem(_items) {
    //items = [{ name: "a", src: "AA", dst: "BBB" }];
    stopAllProgress();
    let parent = $("workbench");
    parent.innerHTML = "";
    for (let i = 0; i < _items.length; i++) {
        let index = i + 1;
        let item = `
        <label class="hide" for=${index}>item</label>
        <input type="checkbox" id="${index}" class="checkbox item-checkbox" onClick="itemCheckedChange(this)" title="${index}"/>
        
        <div class="editor">
            <label for="e${index}" class="inputlabel" >${lightColor(
            _items[i].src.replace(/</g, "&lt;").replace(/\>/g, "&gt;").replace(/\n/g, "\\n")
        )}</label>
            <input type="text" id="e${index}" class="inputbox" value="${_items[i].dst.replace(
            /\n/g,
            "\\n"
        )}" title="${_items[i].name}"/>
    </div>`;
        var div = document.createElement("div");
        div.className = "item";
        div.innerHTML = item;
        parent.appendChild(div);
    }

    // 恢复初始状态
    items = document.getElementsByClassName("item");
    log = $("log");
    checkAll = $("checkall");
    checkAll.checked = false;
    selectAll(checkAll);
    selectCount = 0;
    changeLog("👍完成  ");
}

// 给特定字符加上颜色
function lightColor(str) {
    return str
        .replace(/\$/g, `<span class="high-light">$&</span>`)
        .replace(/\\/g, `<span class="high-light">$&</span>`)
        .replace(/\{/g, `<span class="high-light">$&</span>`)
        .replace(/\}/g, `<span class="high-light">$&</span>`)
        .replace(/\'/g, `<span class="high-light">$&</span>`)
        .replace(/\§/g, `<span class="high-light">$&</span>`)
        .replace(/\(/g, `<span class="high-light">$&</span>`)
        .replace(/\)/g, `<span class="high-light">$&</span>`);
}

// 选择按钮状态改变
function itemCheckedChange(checkbox) {
    if (checkbox.checked) {
        ++selectCount;
    } else {
        --selectCount;
    }
    changeLog(`${selectCount} / ${items.length}`);
    if (selectCount === 0) {
        checkAll.checked = false;
        checkAll.className = "checkbox";
        return;
    }
    // 全选
    if (selectCount === items.length) {
        checkAll.className = "checkbox";
        checkAll.checked = true;
        return;
    }
    // 不全选
    checkAll.checked = false;
    checkAll.className = "checkbox notcheckall";
    return;
}

// 主动翻译
function _translate(button) {
    let source = $("lang-source").value;
    let target = $("lang-target").value;
    if (source === "" || target === "") {
        changeLog("⚠️语言代码不应该为空  ");
        return;
    }
    logProgress("🤔正在翻译  ");
    button.disabled = true;
    setTimeout(() => {
        button.disabled = false;
    }, 3000);
    vscode.postMessage({
        type: "Translate",
        data: `["${source}","${target}"]`,
    });
}
// 用log表示进度
function logProgress(msg) {
    if (progress === undefined) {
        progress = [];
    }
    let count = 0;
    log.innerText = msg;
    let i;
    i = setInterval(() => {
        log.innerText += ".";
        count++;
        if (count > 100) {
            changeLog("❓发生什么事了  ");
        }
    }, 100);
    progress.push(i);
}
// 结束所有log进度
function stopAllProgress() {
    if (progress === undefined) {
        progress = [];
    }
    for (let i = 0; i < progress.length; i++) {
        clearInterval(progress[i]);
    }
}
function changeLog(msg) {
    stopAllProgress();
    log.innerText = msg;
}


src/translate.ts
/*
 * @Author: HumXC Hum-XC@outlook.com
 * @Date: 2022-05-25
 * @LastEditors: HumXC Hum-XC@outlook.com
 * @LastEditTime: 2022-07-26
 * @FilePath: \MCTranslator\src\translate.ts
 * @Description:
 *
 * Copyright (c) 2022 by HumXC Hum-XC@outlook.com, All Rights Reserved.
 */
import {} from "./extension";
import { baidu, checkBaiduError } from "./network";
import { sleep } from "./util";

// 翻译的文档
export type Document = {
    path: string;
    str: string;
    obj: JSON | any;
    doctype: "json" | string;
    // [{翻译的名称,待翻译的原文，翻译结果}]
    result: Array<{ name: string; src: string; dst: string }>;
};
// 由 webview 返回的文档
export type WebviewDocument = {
    path: string;
    str: string;
    obj: JSON | any;
    doctype: "json" | string;
    // [{翻译的名称,待翻译的原文，翻译结果}]
    result: Array<{ name: string; dst: string }>;
};

export async function trHandleBaidu(
    doc: Document,
    source: string,
    target: string,
    appid: string,
    secretkey: string,
    isAdvancedVersion: boolean = false
) {
    let qos = 1;
    let maxLength = 1000;
    let queryStrList: string[] = [];
    if (isAdvancedVersion) {
        qos = 10;
        maxLength = 6000;
    }
    for (let i = 0; i < doc.result.length; i++) {
        queryStrList.push(fomartQueryStr(doc.result[i].src));
    }
    let queryStr = "";
    let index = 0;
    for (let i = 0; i < queryStrList.length; i++) {
        let s = queryStr + queryStrList[i] + "\n";
        let tr = async () => {
            let result = await baidu(queryStr, source, target, appid, secretkey);

            let err = checkBaiduError(result);
            if (err !== undefined) {
                throw new Error(err);
            }

            for (let j = 0; j < result.trans_result.length; j++) {
                let r = result.trans_result[j];
                if (fomartQueryStr(doc.result[j + index].src) === r.src) {
                    doc.result[j + index].dst = deFomartQueryStr(r.dst);
                }
            }

            index = i;
            queryStr = "";
            await sleep(Math.ceil(1000 / qos));
        };

        if (s.length >= maxLength) {
            await tr();
            --i;
            continue;
        } else {
            queryStr = s;
        }
        if (i === queryStrList.length - 1) {
            await tr();
        }
    }
}
function fomartQueryStr(str: string): string {
    if (str === "") {
        return "[%0]";
    }
    // 去除可能会影响签名的字符,转换成自定义的编码
    return str
        .replace(/\n/g, "[%1]")
        .replace(/\\/g, "[%2]")
        .replace(/\&/g, "[%3]")
        .replace(/\+/g, "[%4]")
        .replace(/\§/g, "[%5]")
        .replace(/\#/g, "[%6]")
        .replace(/\_/g, "[%7]");
}
function deFomartQueryStr(str: string): string {
    if (str === "[%0]") {
        return "";
    }
    return str
        .replace(/\[%1\]/g, "\n")
        .replace(/\[%2\]/g, "\\")
        .replace(/\[%3\]/g, "&")
        .replace(/\[%4\]/g, "+")
        .replace(/\[%5\]/g, "§")
        .replace(/\[%6\]/g, "#")
        .replace(/\[%7\]/g, "_");
}
